### DISCUSSION TITLE
Creating custom component in chakra ui

### DISCUSSION THREAD

**Discussion by jude98** [chakra-ui/chakra-ui#8661] [Q&A] [answered]
Hi.
I am trying to create a custom component (`toggle button`) in my project where I am using chakra v2.

I am trying to style the custom toggle using `emotion styled`, where I can pass props and access theme. 
I am not sure if creating a custom component using the emotion styled is the proper way or is there any way of doing it the `chakra way`

Here the issues I am facing is that, 

- cannot access semantic tokens
- it is not typescript compatible with theme such that I cannot get features like autocomplete for theme, even though I have type generated for my theme. 

I am doing something like this:

```javascript
export const ToggleWrapper = styled.button<{
  bg?: string;
  borderradius?: string;
  height?: string;
  width?: string;
}>`
  display: flex;
  align-items: center;
  width: ${({ width }) => width ?? '100%'};
  padding: 1px;
  border: 1px solid;
  border-color: ${({ theme }) => theme.colors.surface.base[500]};
  background: ${({ bg, theme }) => bg || theme.colors.surface.base[200]};
  border-radius: ${({ borderradius }) => borderradius || '32px'};
  cursor: pointer;
  outline: none;
  height: ${({ height }) => height};
`;

export const ToggleElement = styled.span<{
  borderradius?: string;
  fontsize?: string;
  isactive?: boolean;
}>`
  display: flex;
  align-items: center;
  width: 100%;
  padding: 4px 0.5rem;
  border-radius: ${({ borderradius }) => borderradius || '32px'};
  border: 1px solid;
  border-color: ${({ isactive, theme }) =>
    isactive ? theme.colors.brand.secondary[600] : theme.colors.surface.base[500]};
  justify-content: center;
  height: 100%;
  background: ${({ isactive, theme }) =>
    isactive ? theme.colors.brand.secondary[600] : theme.colors.surface.base[200]};
  color: ${({ theme }) => theme.colors.surface.base[200]};
  opacity: ${({ isactive }) => (isactive ? 1 : 0.7)};
  font-size: ${({ fontsize }) => fontsize ?? '1rem'};
  font-weight: 535;
  white-space: nowrap;
  :hover {
    user-select: initial;
  }
`;
```

### Project details:

React + vite + typescript
Chakra version : 2

**Comment by segunadebayo** [comment, marked as answer]
That is v2 plus Emotion `styled`. You lose tokens and typings because `styled` does not go through the Chakra system.

Use the `chakra` factory, or in v3 define a recipe and apply it with `chakra("button", recipe)`.
